### PR TITLE
[apr apr-util] Apr and apr-util for non windows systems

### PR DESCRIPTION
**Pull request by stephen-webb**
The apr and apr-util CMakeLists.txt is Windows only

### CODE REVIEW

**Comment by stephen-webb** [comment]
This PR may address issue #6571 and is a prerequisite for issue #6125

**Review by PhoebeHui — CHANGES_REQUESTED** [review]

**Comment by JackBoosY** [comment]
Needs to rerun test after #9752 merged.

File: triplets/aarch64-linux.cmake [hunk 1]
```diff
@@ -0,0 +1,6 @@
+set(VCPKG_TARGET_ARCHITECTURE aarch64)
+set(VCPKG_CRT_LINKAGE dynamic)
+set(VCPKG_LIBRARY_LINKAGE static)
+
+set(VCPKG_CMAKE_SYSTEM_NAME Linux)
+set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE /opt/xtools/cmake/Toolchain-aarch64-linux.cmake)
```
Comment (PhoebeHui) [inline triplets/aarch64-linux.cmake]: Base on https://github.com/microsoft/vcpkg/blob/master/docs/users/triplets.md#community-triplets

This triplet should move to triplets\community.

File: ports/log4cxx/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL static)
+    file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin ${CURRENT_PACKAGES_DIR}/debug/bin)
+endif()
+
+vcpkg_copy_pdbs()
+
+# Handle copyright
+file(COPY ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/log4cxx)
+file(RENAME ${CURRENT_PACKAGES_DIR}/share/log4cxx/LICENSE ${CURRENT_PACKAGES_DIR}/share/log4cxx/copyright)
```
Comment (PhoebeHui) [inline ports/log4cxx/portfile.cmake]: Could you please update it to 

file(Install ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)

File: ports/log4cxx/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,31 @@
+include(vcpkg_common_functions)
```
Comment (PhoebeHui) [inline ports/log4cxx/portfile.cmake]: Could you remove it? this is deprecated.

File: ports/apr/portfile.cmake [hunk 4]
```diff
@@ -17,46 +17,80 @@ vcpkg_extract_source_archive_ex(
...
+    vcpkg_execute_required_process(
+        COMMAND make install
+        WORKING_DIRECTORY ${SOURCE_PATH}
+        LOGNAME install-${TARGET_TRIPLET}-release
+    )
 endif()
 
 # Handle copyright
 file(COPY ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/apr)
 file(RENAME ${CURRENT_PACKAGES_DIR}/share/apr/LICENSE ${CURRENT_PACKAGES_DIR}/share/apr/copyright)
```
Comment (PhoebeHui) [inline ports/apr/portfile.cmake]: Could you update it to 

file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)?

File: ports/apr-util-windows/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,46 @@
+include(vcpkg_common_functions)
```
Comment (PhoebeHui) [inline ports/apr-util-windows/portfile.cmake]: Could you please remove it? this is deprecated.

**Review by PhoebeHui — CHANGES_REQUESTED** [review]

**Comment by PhoebeHui** [comment]
@stephen-webb, thanks for updates! 

The CI test failed on windows, linux and OSX, could you take a look?

**Comment by stephen-webb** [comment]
The CI fails because apr-util-unix is being built by the Windows systems and apr-util-windows is being built by the Linux and OSX systems.

I have moved apr-util-unix and apr-util-windows back into apr-util and used a conditional inside ports.cmake

File: ports/apr-util/portfile.cmake [hunk 6]
```diff
@@ -1,46 +1,96 @@
...
-  set(APU_DECLARE_EXPORT ON)
-  set(APU_DECLARE_STATIC OFF)
-else()
-  set(APU_DECLARE_EXPORT OFF)
-  set(APU_DECLARE_STATIC ON)
-endif()
+if(VCPKG_TARGET_IS_WINDOWS)
+
+    vcpkg_download_distfile(ARCHIVE
+      URLS "https://archive.apache.org/dist/apr/apr-util-1.6.0-win32-src.zip"
```
Comment (JackBoosY) [inline ports/apr-util/portfile.cmake]: I found that the version used in windows is not the same as the version used in unix, we need to use the same version.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by stephen-webb** [comment]
Most CI build failures are due to openMAMA failing. APR and APR-UTIL builds succeeded except x64-windows-static which reported:
2020-02-11T03:52:38.2175510Z -- Downloading https://www.apache.org/dist/apr/apr-1.7.0.tar.bz2...
2020-02-11T03:53:08.7265036Z -- Downloading https://www.apache.org/dist/apr/apr-1.7.0.tar.bz2... Failed. Status: 35;"SSL connect error"

The failure must be network related as x86-windows did not have a download issue. It reported:
2020-02-11T03:51:55.7940655Z -- Downloading https://www.apache.org/dist/apr/apr-1.7.0.tar.bz2...
2020-02-11T03:52:22.9621198Z -- Extracting source C:/vsts/_work/3/s/downloads/apr-1.7.0.tar.bz2

**Comment by JackBoosY** [comment]
@stephen-webb I also cannot use this link to download the source code on my machine. Can you replace the source url?

**Comment by stephen-webb** [comment]
Which link do you think needs to be changed?
Note: All links are the official release archives.

**Comment by JackBoosY** [comment]
@stephen-webb Forgot it. I will restart ci test after #10051 merged.

**Comment by JackBoosY** [comment]
Seems have build failures, can you resolve them?

Thanks.

**Comment by stephen-webb** [comment]
The failure do not relate to my changes - they are from other ports.

**Comment by JackBoosY** [comment]
Still have some regressions.

**Comment by stephen-webb** [comment]
Once again, the failures are in other ports! activemq-cpp, aliyun-oss-c-sd, openmama

**Comment by JackBoosY** [comment]
@stephen-webb Sorry, I forgot that.
Please remove `apr:x64-linux=fail` and `apr:x64-osx=fail`, and add the following code to
 _VCPKG_PATH/scripts/ci.baseline.txt_:
```
activemq-cpp:x64-linux=fail
activemq-cpp:x64-osx=fail
aliyun-oss-c-sdk:x64-linux=fail
aliyun-oss-c-sdk:x64-osx=fail
openmama:x64-linux=fail
openmama:x64-osx=fail
```

Thanks.

File: ports/apr-util/portfile.cmake [hunk 7]
```diff
@@ -1,46 +1,92 @@
...
+    # set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE ${MY_CROSS_DIR}/cmake/Toolchain-${VCPKG_PLATFORM_TOOLSET}.cmake)
+    # set(CONFIGURE_PARAMETER_1 --host=${VCPKG_PLATFORM_TOOLSET})
+    # set(CONFIGURE_PARAMETER_2 --cache-file=${MY_CROSS_DIR}/autoconf/${VCPKG_PLATFORM_TOOLSET}.cache)
+    if(CONFIGURE_PARAMETER_1)
+        message(STATUS "Configuring apr-util with ${CONFIGURE_PARAMETER_1} ${CONFIGURE_PARAMETER_2} ${CONFIGURE_PARAMETER_3}")
+    else()
+        message(STATUS "Configuring apr-util")
+    endif()
+
+    vcpkg_execute_required_process(
```
Comment (JackBoosY) [inline ports/apr-util/portfile.cmake]: We already have functions `vcpkg_*_make` to handle this situation, could you please replace these lines with those functions?
See documentation:
[vcpkg_configure_make](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_configure_make.md)
[vcpkg_install_make](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_install_make.md)

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/apr-util/portfile.cmake (hunk 7 above)
Comment (JackBoosY) [reply ports/apr-util/portfile.cmake]: Example: [sdl1](https://github.com/microsoft/vcpkg/blob/bf75bdfe92206e0ac5461743b09592495137c674/ports/sdl1/portfile.cmake)

**Comment by stephen-webb** [comment]
CI fails the static build because it installs apr-1-config in bin, which APR-Util need to find to find APR.

How do I tell CI to ignore/skip static build?

**Comment by JackBoosY** [comment]
@stephen-webb Just add `apr:x64-linux=fail` and `apr:x64-osx=fail` to _VCPKG_PATH/scripts/ci.baseline.txt_ would be okay.

**Comment by stephen-webb** [comment]
Now libpq and spdk-ipsec are failing!

**Comment by JackBoosY** [comment]
It seems that there are other issues with the CI system, and after resolving the issue, I will restart this PR test.

**Comment by stephen-webb** [comment]
libpq still failing

**Comment by PhoebeHui** [comment]
@stephen-webb, we will rerun the CI testing after PR https://github.com/microsoft/vcpkg/pull/10445 merged.

File: ports/apr/portfile.cmake [hunk 8]
```diff
@@ -4,59 +4,85 @@ endif()
...
+        PREFER_NINJA
+        OPTIONS
+            -DINSTALL_PDB=OFF
+            -DMIN_WINDOWS_VER=Windows7
+            -DAPR_HAVE_IPV6=ON
+            -DAPR_INSTALL_PRIVATE_H=${INSTALL_PRIVATE_H}
+            ${FEATURE_OPTIONS}
+        # OPTIONS -DUSE_THIS_IN_ALL_BUILDS=1 -DUSE_THIS_TOO=2
+        # OPTIONS_RELEASE -DOPTIMIZE=1
+        # OPTIONS_DEBUG -DDEBUGGABLE=1
```
Comment (PhoebeHui) [inline ports/apr/portfile.cmake]: Could you remove them if they doesn't need any more?

File: ports/apr/portfile.cmake [hunk 9]
```diff
@@ -4,59 +4,85 @@ endif()
 
 include(vcpkg_common_functions)
```
Comment (PhoebeHui) [inline ports/apr/portfile.cmake]: Could you also remove 'include(vcpkg_common_functions)', this is deprecated.

**Review by PhoebeHui — CHANGES_REQUESTED** [review]

**Comment by JackBoosY** [comment]
This PR LGTM, we just need to wait for ci to fix.

**Review by JackBoosY — APPROVED** [review]

**Review by PhoebeHui — APPROVED** [review]

**Comment by JackBoosY** [comment]
LGTM.

**Comment by strega-nil** [comment]
Alright, let's merge!

**Pull request merged**